src/cycle_basis.c

```diff
@@ -0,0 +1,660 @@
+                const igraph_vector_t *weights,
+                igraph_vector_ptr_t *basis,
+		) {
+
+    /* General plan:
+     *
+     * 0. Find a feedback vertex set Z: all cycles pass through a vertex
+     *    in Z. Worst case Z := {all vertices}. Basically, 2-degree nodes
+     *    are excluded because there's no decision: the path goes in one
+     *    way and out another. Of course, 0-degree are also not needed.
+     * 1. Construct all shortest path trees 
+     * 2. Construct candidate set of cycles, sorted in nondecreasing
+     *    order of weight. Each cycle is attached to a tree (but more than
+     *    one cycle can be attached to the same tree).
+     * 3. for i = 1 to N do
+     *     3.1 compute vector Si s.t. <Cj, Si> = 0 for each 1 <= j < i
+     *         (orthogonal to the current cycles)
+     *     3.2 for all trees, update vertex labels based on Si
+     *     3.3 for each candidate cycle, compute <C, Si> using the labels
+     *     3.4 append the least-weight cycle non-orthogonal to Si
+     *         (so we can normalize it into <C, Si> = 1
+     *
+     * */
+    /* TODO: finish implementation */
+
+    long int i, n;
+    long int no_of_nodes = igraph_vcount(graph);
+    igraph_vector_t si;
+    igraph_vector_int_t nonorth_cycles;
+
+    igraph_vector_t fvs, basis_weight;
+    /* list of vertices = tree -> list of trees */
+    igraph_vector_ptr_t trees;
+    /* NOTE: [tree_id, edge_id, tree_id2, edge_id2, ...] */
+    igraph_vector_int_t candidate_cycles;
+    /* total weight of each cycle */
+    igraph_vector_t candidate_weights;
+
+    /* 0. Feedback vertex set */
+    IGRAPH_CHECK(igraph_i_feedback_vertex_set(
+			    graph, &fvs, weights));
+    n = igraph_vector_size(&fvs);
+
+    /* 1. Construct all shortest path trees */
+    IGRAPH_CHECK(igraph_i_shortest_path_trees(
+			    graph, weights, &fvs, &trees));
+
+    /* 2. Construct candidate list */
+    IGRAPH_CHECK(igraph_i_candidate_cycles(
+			    graph, weights,
+			    &candidate_cycles, &candidate_weights));
+
+    /* 3. Construct basis one cycle at a time
+     *    similar to a standard Gram Schmidt orthogonalization */
+    for (i = 0; i < no_of_nodes; i++) {
+        /* 3.1 compute Si, the nonnormalized, minimal vector
+	 *     sticking out from the current space */
+	IGRAPH_CHECK(igraph_i_compute_Si(graph, weights, basis, i, &si));
+
+	/* 3.2 update tree vertex labels from Si */
+	IGRAPH_CHECK(igraph_i_update_trees_vertex_labels(
+			&trees, &si));
+
+	/* 3.3 find non-orthogonal candidates */
+	IGRAPH_CHECK(igraph_i_nonorthogonal_candidates(
+			candidate_cycles, &si, &nonorth_cycles));
+
+	/* 3.4 get the shortest nonorthogonal cycle */
+	IGRAPH_CHECK(igraph_i_shortest_nonorthogonal_cycle(
+			candidate_cycles, &candidate_weights, &nonorth_cycles, basis));
+
+    }
+
+
+    /* Clean */
+    igraph_vector_destroy(&fvs);
+    igraph_vector_ptr_destroy_all(trees);
+    IGRAPH_FINALLY_CLEAN(1 + n + 1); // FIXME: Check this number
```

Comment:
This may be a useful read:

https://igraph.discourse.group/t/who-should-clean-up-vector-ptr-out-argument-in-case-of-interruption-or-error/153

Unfortunately, the handling of vector_ptrs in an error conditions is not completely standardized yet.

I am not aware of any existing case where all vector_ptr elements are put on the finally stack, and maybe it would be better not to start doing this. See the above thread for how this is usually handled.

Something to pay attention to:

Should the function exit due to an error condition *or* interruption, it should be safe for the caller to destroy/free any non-NULL elements of the ptr_vector. Thus, anything you destroy must be set to NULL. Alternatively, if everything was destroyed, the vector_ptr can be cleared.

As far as I can tell, existing functions do not leave any elements undestroyed or unfreed in an error condition.

Tip: `igraph_Free(ptr)` will set `ptr` to `NULL`, but `igraph_free` will not (cannot).

Reply:
> I am not aware of any existing case where all vector_ptr elements are put on the finally stack

The size of the finally stack is limited so probably not all elements would fit there anyway. So far I was following the principle that the vector_ptr takes "ownership" of any pointer that you put into it, so you only need to put a cleanup function on the finally stack that iterates over the elements of the vector_ptr and destroys and frees the non-null ones. This function can then be popped off the finally stack when the vector_ptr is returned to the user. The "item destructor" pattern outlined in the Discourse group is supposed to help with the implementation.